sln-add: Support for slnx

edvilme opened the pull request:
Contributes to #40913 
> The dotnet CLI should support the new slnx format for building and in the existing solution management commands. It should also help interested users migrate to the new format.

This adds `dotnet sln add` support for .slnx files

edvilme committed:
[IMP] sln-list: Support for slnx

Refactor code

Refactor code

Add translations

Fix typo

Fix issues with --solution-folders

Standarize error messages and tests

Fix solution folder formatting

Nit: Address pr comments

edvilme committed:
sln-add: Support for slnx

edvilme committed:
Handle solution folders

edvilme committed:
Fix tests

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -42,40 +44,41 @@ public AddProjectToSolutionCommand(ParseResult parseResult) : base(parseResult)
+        {
+            ISolutionSerializer serializer = SlnCommandParser.GetSolutionSerializer(solutionFileFullPath);
+            SolutionModel solution = await serializer.OpenAsync(solutionFileFullPath, cancellationToken);
+            foreach (var projectPath in projectPaths)
```
oh great, thx!

edvilme committed:
Fix UTF8 BOM tests

edvilme committed:
Catch errors

edvilme committed:
Fix additional tests

edvilme committed:
Fix additional tests

edvilme committed:
Fix additional issues

edvilme committed:
Remove sdk.slnx file

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,109 @@
         {
-            return Path.GetDirectoryName(path);
+            // SolutionModel::AddFolder expects path to have leading, trailing and inner forward slashes
+            return PathUtility.EnsureTrailingForwardSlash( PathUtility.GetPathWithForwardSlashes(Path.Join(" / ", _solutionFolderPath)) );
```
Yes, `Path.Join("/", _solutionFolderPath)` adds the leading forward slash :)

edvilme committed:
Fix additional tests

edvilme committed:
Fix additional tests

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,125 @@
+                }
+                catch (Exception ex)
+                {
+                    // TODO: Update with error codes from vs-solutionpersistence
+                    // Duplicate project
+                    if (Regex.Match(ex.Message, @"Project name '.*' already exists in the solution folder\.").Success || Regex.Match(ex.Message, @"Duplicate item '.*' of type 'Project'\.").Success)
+                    {
+                        Reporter.Output.WriteLine(
+                            string.Format(CommonLocalizableStrings.SolutionAlreadyContainsProject, solutionFileFullPath, relativePath));
+                        continue;
+                    }
```
This is needed at the moment because the current behavior expects the command to succeed if a duplicate project is to be added, while vs-solutionpersistence throws an error.  This attempts to catch that error. 
In the future, there will be better ways of identifying these exceptions

edvilme replied:
Oh didn't see that, will take a look :)

edvilme committed:
Fix additional tests

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,125 @@
-                return _relativeRootSolutionFolders;
+                // Get full project path
+                var relativePath = Path.GetRelativePath(Path.GetDirectoryName(solutionFileFullPath), projectPath);
+                try
```
I like that except that if we do that, it would cause the whole method to throw, which might be inconvenient if there are several projects that we want to add there

edvilme committed:
Fix duplicate project tests

edvilme committed:
Fix tests

edvilme committed:
108/133 tests passing

edvilme committed:
Work on tests

edvilme committed:
Nit

edvilme committed:
Refactored code to fix guid tests

edvilme committed:
Fix tests

baronfel commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,145 @@
+            SolutionProjectModel project;
+            try
+            {
+                solution.AddProject(relativePath, guid, solutionFolder);
```
I'm not so sure @kasperk81 - the library doesn't load the project and ask it for its ProjectTypeGuid, which means that either
* we artificially limit the project types we support (which would be a feature regression)
* we catch the exception, then re-parse and add the project type to the parsing library's ProjectTable (wasteful)
* we set up the project table ahead of time (why bother)?

In this scenario we are more knowledegable and extensible than the library is about the kinds of projects users want to add to the solution, so we should keep our existing interaction pattern.

edvilme replied:
> Do we actually need to handle this? It seems like a corner-case error that isn’t essential for us to manage. For example, if someone runs something like:
> 
> ```shell
> echo "foobar" > foo.csproj && dotnet sln add foo.csproj
> ```
> 
> ...we could just let them proceed. The `dotnet build` command will ultimately validate the project, so it doesn’t seem like this validation needs to be a concern for `dotnet sln add` in the first place.

Yes, I would generally agree with you. Yet, I think the scope here is to implement the new library with as few user-facing changes as possible. The fact that we have 2 tests on this to me signifies users expect this behavior :)

edvilme replied:
I definitely agree on both points
Just did some testing and reverted back the changes for guid detection.
On the SlnFile on tests, I 100% agree. However I want to get _most_ of the tests right before changing anything to them. At most, I have changed string matchings. But that is definitely the plan :)

edvilme committed:
Fix some config tests

edvilme committed:
Revert guid detection

nagilson commented on src/Cli/dotnet/commands/dotnet-sln/LocalizableStrings.resx:
```diff
@@ -184,9 +184,9 @@
   </data>
   <data name="SerializerNotFound" xml:space="preserve">
-    <value>Could not read solution file {0}. Supported files are .sln and .slnx valid solutions.</value>
+    <value>Could not find serializer for file {0}.</value>
```
The new strings are valid too, but probably the original strings are clearer, so they are preferred. If you disagree, let me know :) 

You don't need to fix the history, you could just change this in a new commit to what it previously was so that it doesn't get changed.

edvilme replied:
Missed this comment, I will include the updated translations in my next commit. Thanks both! :D

edvilme commented:
Turns out the heuristics for generating project GUIDs are different between the old and new serializers, which was causing several tests to break. Replacing them should have no repercussions as they are mostly for identifying the projects individually

edvilme committed:
Update tests guids and translations

edvilme committed:
Update translations (build)

edvilme committed:
Fix additional tests

edvilme committed:
Fix issues from pr

edvilme committed:
Solve solution folder tests

edvilme committed:
Fix additional tests

edvilme committed:
Refactor code

nagilson commented:
> running `dotnet sln add foo.weirdproj` currently results in:
> 
> * `project 'foo.weirdproj' has an unknown project type and cannot be added to the solution file. contact your sdk provider for support.` — when the xml is valid
> * `invalid project 'foo.weirdproj'. the project file could not be loaded. data at the root level is invalid. line 1, position 1.` — when the xml is invalid
> 
> the second error reflects validation behavior that goes against the single responsibility principle, as `dotnet sln add` shouldn’t be responsible for project file validation (unlike `dotnet build`).
> 
> tests conforming to `SlnFile` behavior (that passing an explicit unknown guid should succeed) have no user-facing meaning because, fortunately, there is no guid argument in the `dotnet sln add` command. removing these tests and related special error handling would simplify the code.
> 
> if tools like vs, vscode, slngen, and the sdk are transitioning to `vs-solutionpersistence`, there’s no reason to retain the discontinued monodevelop's `SlnFile` behavior.
> 
> these are all the “error” cases blocking this pr for two weeks, but none of them are actually relied upon for functionality. a note in the 9.0.2x docs like “we've switched to a new solution file library from the core vs team, replacing the old one from the discontinued monodevelop” should suffice to explain minor (git-diff visible) differences without impacting functionality.

I think you raise good points. @edvilme chatted with me regarding this, and we think we can remove these error cases blocking this PR. We are also going to go over these error cases with the SLNX team to make sure that none of these old behaviors are behaviors they deem to be important to keep.

edvilme committed:
Update tests

edvilme committed:
Fix tests

Most tests are now fixed. The rest are still pending due to design considerations and coordination with vs-solutionpersistence team

edvilme committed:
Fix WhenProjectWithAdditionalConfigurationsIsAddedSolutionDoesNotMapThem

edvilme committed:
Revert changed project guids

Guids are now read from project info

edvilme committed:
Revert changed project guids

Guids are now read from project info

edvilme committed:
Fix project config tests

edvilme committed:
Fix tests

edvilme committed:
Fix all tests + Update vs-solutionpersistence

edvilme committed:
Nit

edvilme committed:
Nit

edvilme committed:
Nit

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
-        {
-            if (_inRoot)
+            // Set default configurations and platforms for sln file
+            foreach (var platform in new List<string>{ "Any CPU", "x64", "x86" })
```
This was included in the previous implementation in `SlnFileExtensions` (https://github.com/dotnet/sdk/blob/e6c8e57159724c0d22640f5eadcd32954276b22f/src/Cli/dotnet/SlnFileExtensions.cs#L128C9-L151C10) but we could potentially make these a constant in case we want to modify them in the future

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
-            return !path.StartsWith("..");
-        }
+            // Open project instance to see if it is a valid project
+            ProjectRootElement projectRootElement = ProjectRootElement.Open(relativePath);
```
This is needed: 
- to verify that the project file is valid. This is outside the scope of vs-solutionpersistence, and needed to preserve existing behavior
- as a fallback for getting the project type guid from the file directly in cases where it is not recognized by vs-solutionpersistence

edvilme committed:
Fix whitespaces

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
+            foreach (var solutionPlatform in solution.Platforms)
+            {
+                var projectPlatform = projectInstancePlatforms.FirstOrDefault(
+                    x => x.Replace(" ", string.Empty) ==  solutionPlatform.Replace(" ", string.Empty), projectInstancePlatforms.FirstOrDefault("Any CPU"));
```
The current behavior expects to match platform regardless of whitespaces, so `AnyCPU == AnyCPU`. For the majority of cases this shouldn't have additional consequences.

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
+            foreach (var solutionBuildType in solution.BuildTypes)
+            {
+                var projectBuildType = projectInstanceBuildTypes.FirstOrDefault(
+                    x => x.Replace(" ", string.Empty) == solutionBuildType.Replace(" ", string.Empty), projectInstanceBuildTypes.FirstOrDefault("Debug"));
```
This exists as the current behavior, however this is more of a different implementation on our end.

edvilme reviewed (commented):
All tests passing now, but leaving a few comments for future reference :)

edvilme committed:
[TEST] sln-add: Add example slnx files, and parameters

edvilme committed:
[TEST] sln-add: Use vs-solutionpersistence on templates

edvilme committed:
[TEST] sln-add: Add SolutionFilesTemplates

edvilme committed:
[TEST] sln-add: Compare slnx templates

edvilme committed:
[TEST] sln-add: Fix sln-templates tests

edvilme committed:
[TEST] sln-add: Migrate all tests

edvilme committed:
Fix typo

edvilme commented on test/dotnet-sln.Tests/GivenDotnetSlnAdd.cs:
Here, const strings for expected sln results were moved to a folder under TestAssets, with their .slnx counterparts generated by `dotnet sln migrate`.  
Additional test cases were created for sln.
`SlnFile` was removed from these tests as well

edvilme committed:
[TEST] sln-list: Update testAsset identifiers

edvilme commented:
> the identifier doesn't help because testmanager's GetTestDestinationDirectoryPath shortens the directory name to 24 characters and we end up having the same errors.

Yes, it pushed before I was able to see your pr, bit of a sync issue there... I like the idea on your PR as it allows us to have a greater variety of different names for different cases (of which there are many on this pr)

edvilme committed:
[TEST] restore: Add App.sln to argument list

edvilme committed:
Merge branch 'main' into edvilme-slnx-add

edvilme commented:
> getting closer! 👍
> 
> these tests are failing:

Hii, thanks and thanks for your input. Yes, those tests are currently failing on several PRs and should be unrelated to the changes here. When they're fixed, I will update the branch :)

edvilme committed:
Merge branch 'main' into edvilme-slnx-add

edvilme commented:
You're right. There were some other tests from templating failing, so I got them mixed up. The issue here is when adding projects from a different directory, e.g., `dotnet sln subdirectory/App.sln add subdirectory/Project.csproj`. I will look into that

edvilme committed:
sln-add: When working and solution directories are different, resolve paths correctly

edvilme committed:
Merge branch 'edvilme-slnx-add' of https://github.com/edvilme/sdk into edvilme-slnx-add

kasperk81 reviewed (approved):
lgtm (one nit)

edvilme committed:
Nit

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
-            PathUtility.EnsureAllPathsExist(arguments, CommonLocalizableStrings.CouldNotFindProjectOrDirectory, true);
-
-            var fullProjectPaths = _arguments.Select(p =>
+            try
```
nit: \n before this

(Also just anywhere you have }, I prefer a \n before the next thing with few exceptions)

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
+                AddProjectsToSolutionAsync(solutionFileFullPath, fullProjectPaths, CancellationToken.None).Wait();
+                return 0;
+            }
+            catch (GracefulException)
```
The other catch explicitly excludes this case, and this case just throws, so we can skip this block.

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
-
-            var currentDirString = $".{Path.DirectorySeparatorChar}";
-            if (projectFilePath.StartsWith(currentDirString))
+            ISolutionSerializer serializer = SlnCommandParser.GetSolutionSerializer(solutionFileFullPath);
```
Is this a new runtime concept? I don't see it in the SDK currently or in this PR, and a quick bing (google) search didn't reveal anything interesting. If so, should we put this behind #ifdefs?

edvilme replied:
Which part are you refering to here?

edvilme replied:
Oh yes. It is a new serializer for sln(x) solution files that was recently released as open source and is being used by VS and MSBuild. 
https://github.com/microsoft/vs-solutionpersistence

edvilme replied:
The goal long term is to replace `SlnFile` with this

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
+            ISolutionSerializer serializer = SlnCommandParser.GetSolutionSerializer(solutionFileFullPath);
+            SolutionModel solution = await serializer.OpenAsync(solutionFileFullPath, cancellationToken);
+            // set UTF8 BOM encoding for .sln
+            if (serializer is ISolutionSerializer<SlnV12SerializerSettings> v12Serializer)
```
What's the significance of V12?

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
-        {
-            if (_inRoot)
+            // Set default configurations and platforms for sln file
+            foreach (var platform in new[]{ "Any CPU", "x64", "x86" })
```
I thought we supported ARM now? Not relevant here?

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
             }
-
-            if (_relativeRootSolutionFolders != null)
+            foreach (var buildType in new []{ "Debug", "Release" })
```
Most of my work is halfway between these: Rebug

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
             {
-                // The user has specified an explicit root
-                return _relativeRootSolutionFolders;
+                solution.AddBuildType(buildType);
```
Should we be detecting what the solution currently does for any of these? I'm wondering if it could turn, for instance, a debug-only slnx into debug|release

edvilme replied:
I believe it mainly serves for mapping with the projects' build type

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
-                PathUtility.EnsureTrailingSlash(slnFile.BaseDirectory),
-                fullProjectPath);
+            SolutionFolderModel? solutionFolder = (!_inRoot && _solutionFolderPath != null)
+                ? solution.AddFolder(GetSolutionFolderPathWithForwardSlashes(_solutionFolderPath))
```
What does this do if that already exists?

edvilme replied:
If a solution folder already exists, said folder is returned 
https://github.com/microsoft/vs-solutionpersistence/blob/87ee8ea069662d55c336a9bd68fe4851d0384fa5/src/Microsoft.VisualStudio.SolutionPersistence/Model/SolutionModel.cs#L183C9-L187C10

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
+                }
+                catch (InvalidProjectFileException ex)
+                {
+                    Reporter.Error.WriteLine(string.Format(CommonLocalizableStrings.InvalidProjectWithExceptionMessage, projectPath, ex.Message));
```
These are nonfatal? They're nonfatal because another project might have been successfully added?

edvilme replied:
These are non-fatal as to comply with current behavior and because there are tests expecting this. But we could further discuss this if needed:)

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
+                if (!string.IsNullOrEmpty(relativePathDirectory))
+                {
+                    SolutionFolderModel relativeSolutionFolder = solution.AddFolder(GetSolutionFolderPathWithForwardSlashes(relativePathDirectory));
+                    project.MoveToFolder(relativeSolutionFolder);
```
Is it possible for there to be multiple folders in this hierarchy?

edvilme replied:
You mean nested folders? If that is the case, yes, vs-solutionpersistence adds intermediate folders as needed

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
+            foreach (var solutionPlatform in solution.Platforms)
+            {
+                var projectPlatform = projectInstancePlatforms.FirstOrDefault(
+                    x => x.Replace(" ", string.Empty) ==  solutionPlatform.Replace(" ", string.Empty), projectInstancePlatforms.FirstOrDefault("Any CPU"));
```
nit: better name for 'x'?

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
-        }
+            foreach (var solutionPlatform in solution.Platforms)
+            {
+                var projectPlatform = projectInstancePlatforms.FirstOrDefault(
```
If there are multiple, we only care about the first?

edvilme replied:
I agree. Here I tried to avoid breaking any tests and changing the behavior of the command as to not impact users. 
Here is the current test: https://github.com/dotnet/sdk/blob/e74e404c24adbcea8468529191a950d86c5473c7/test/dotnet-sln.Tests/GivenDotnetSlnAdd.cs#L1153

edvilme replied:
I also think if it needs to map to a build type/platform that does not match the ones in the solution, it makes sense to default to the first one

Forgind commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,183 @@
+            foreach (var solutionBuildType in solution.BuildTypes)
+            {
+                var projectBuildType = projectInstanceBuildTypes.FirstOrDefault(
+                    x => x.Replace(" ", string.Empty) == solutionBuildType.Replace(" ", string.Empty), projectInstanceBuildTypes.FirstOrDefault("Debug"));
```
When would we have a space in Debug or Release? Also, should we be using case-insensitive matches here?

edvilme replied:
Ideally we shouldn't, this exists as to preserve existing behavior. This was in a test that matched "FooBar" and "Foo Bar"

Forgind reviewed (commented):
Mostly just questions 🙂 Thanks for all your work to enable this!

edvilme committed:
Nit

edvilme committed:
Nit

nagilson reviewed (commented):
Many questions arise, but this is looking good :) I haven't looked at the tests in their entirety -- its possible some of my questions are answered in there. It would be ideal to have this looked at by the sln persistence team as well but I think once all of my comments are resolved and I looked over the tests I can approve it! :fireworks:

nagilson reviewed (commented):
The review tool didnt save some of my comments so I have to resubmit them, sorry!

nagilson reviewed (commented):
(Trying to submit the comments again through the tool -- might not work... (I had to resubmit them individually 🙄 ))

nagilson commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,180 @@
+                {
+                    Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true)
+                });
             }
```
Is this the expected behavior? With the old code, I tried to do dotnet sln add on a project that was a release project where the solution only had a debug configuration (I removed the Release line) and it did not add a Release line to the solution.

edvilme replied:
You mean the default configurations or the encoding? If it is the former, yes, there is a test that explicitly expects UTF8 encoding fo sln files. If it is the latter, I can try experimenting a bit with this, as it indeed seems weird

nagilson replied:
It's about the configurations!

nagilson replied:
⭐  I think this is the only important remaining thing now.. This is more about the 
` // Set default configurations and platforms for sln file` line. 

Basically, when I had tested it: if you do `dotnet new console -n release`, and `dotnet new sln`, then remove the release line from the sln and `dotnet sln add release`, I think it didn't used to have a release line after add, but now I think it does. Is that ok? ... Maybe? It is a 'breaking change' if you can replicate that behavior. I would consult with the sln team if so on if this is ok.

edvilme replied:
Oh. Removing this does cause some tests to break. But after analyzing it further, it might be by how the sln files it compares them to are generated. I will try replicating it and see what it generates :)

edvilme replied:
I looked into this and this is required for some tests to pass but we could potentially bring this to discussion if we want to add arm64 support by default @nagilson @baronfel

edvilme replied:
The current behavior also adds these default configurations, https://github.com/dotnet/sdk/blob/10d457d2b369ea1fbcb8c3f6d02bd43e08411146/src/Cli/dotnet/SlnFileExtensions.cs#L128
So this shouldn't be a breaking issue :)

nagilson commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,180 @@
-                return _relativeRootSolutionFolders;
+                var relativePathDirectory = Path.GetDirectoryName(pathRelativeToSolutionFile);
+                if (!string.IsNullOrEmpty(relativePathDirectory))
+                {
```
In the old behavior, you could do dotnet sln add -s ..\foo r. It would add .. to the folder as a dir and then put r/ in r/r.csproj if there was a duplicate project not called r in the folder foo. Now, the API fails: One or more errors occurred. (Names cannot:

contain any of the following characters: / ? : \ * " < > |
contain control characters
be system reserved names, including 'CON', 'AUX', 'PRN', 'COM1' or 'LPT2'
be '.' or '..' (Parameter 'name'))
Is this an expected breaking change?

edvilme replied:
The new API does have a method that validates names for both projects and solution folders. The restrictions make sense to me considering OS' rules, but if it did work on the old behavior, it is possible this _is_ a breaking change

nagilson replied:
Ok, awesome. That's not a blocker (I mean its not even within our control) but we should mark this as a breaking change and it likely needs a doc, if that hasn't been said already. I don't know if youre keeping a list of stuff but I would recommend you do.

nagilson commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,180 @@
+                    }
+                }                
             }
+            // Add settings based on existing project instance
```
Note to self: We dont need to specify the global properties since the sln doesn't manage them and it doesn't seem to impact anything.

nagilson commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,180 @@
-            return Path.GetDirectoryName(path);
+            foreach (var solutionBuildType in solution.BuildTypes)
+            {
+                var projectBuildType = projectInstanceBuildTypes.FirstOrDefault(
```
Is debug what we want for the default behavior? Is that documented anywhere?

edvilme replied:
You are right. This is repetitive as we are adding the default build types anyways and "Debug" is the first one

nagilson replied:
Good point, thank you.

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,180 @@
+        private static string GetSolutionFolderPathWithForwardSlashes(string path)
+        {
+            // SolutionModel::AddFolder expects paths to have leading, trailing and inner forward slashes
+            return "/" + string.Join("/", PathUtility.GetPathWithDirectorySeparator(path).Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries)) + "/";
```
Best we have right now is a comment on their code https://github.com/microsoft/vs-solutionpersistence/blob/87ee8ea069662d55c336a9bd68fe4851d0384fa5/src/Microsoft.VisualStudio.SolutionPersistence/Model/SolutionModel.cs#L171C1-L172C1

nagilson commented on test/dotnet-sln.Tests/GivenDotnetSlnAdd.cs:
```diff
@@ -1346,108 +1058,150 @@ public void WhenSolutionFolderAndInRootIsPassedItFails(string solutionCommand)
                 .WithSource()
                 .Path;
 
+            // TODO: Move to different location
```
nit: do we want to do this? I dont think it is needed.

edvilme replied:
Doesn't feel very clean to remove the sln(x) files. But if people are ok with it, I am happy to leave it as to avoid generating even more test assets unnecessarily

nagilson commented on test/dotnet-sln.Tests/GivenDotnetSlnAdd.cs:
```diff
@@ -29,384 +33,6 @@ public GivenDotnetSlnAdd(ITestOutputHelper log) : base(log)
-		Debug|x86 = Debug|x86
-		Release|Any CPU = Release|Any CPU
-		Release|x64 = Release|x64
-		Release|x86 = Release|x86
```
Great idea to move all of these to its own file.

edvilme replied:
Haha thx, it was getting huge (even more so with the slnx variations)

nagilson commented:
Ok, I just finished looking through the tests as well actually so once all of the feedback is accounted for I am happy to merge (still would be nice to have the solutionpersistence review)

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,180 @@
-
-            SlnArgumentValidator.ParseAndValidateArguments(_fileOrDirectory, _arguments, SlnArgumentValidator.CommandType.Add, _inRoot, relativeRoot);
+            _solutionFolderPath = parseResult.GetValue(SlnAddParser.SolutionFolderOption);
+            SlnArgumentValidator.ParseAndValidateArguments(_fileOrDirectory, _projects, SlnArgumentValidator.CommandType.Add, _inRoot, _solutionFolderPath);
```
Kindof. `isRelativeRoot` was only used to check if the passed --solution-folder option was not null. This behavior is replicated further down 
```cs
SolutionFolderModel? solutionFolder = (!_inRoot && _solutionFolderPath != null)
    ? solution.AddFolder(GetSolutionFolderPathWithForwardSlashes(_solutionFolderPath))
    : null;
```

(I should probably replace it with `!string.IsNullOrEmpty` now that I think about it :)

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,180 @@
 
+        private static string GetSolutionFolderPathWithForwardSlashes(string path)
+        {
+            // SolutionModel::AddFolder expects paths to have leading, trailing and inner forward slashes
```
I am not sure if this answers the question but `AddFolder` does expect `/` and would throw if `\` is passed. Likewise the `.RemoveEmptyEntries` is there as some of the tests originally expected double forward slashes to be ignored

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,180 @@
             }
-
-            if (_relativeRootSolutionFolders != null)
+            // Generate intermediate solution folders
```
Oh good catch, didn't notice. I don't mind removing it

edvilme committed:
Nit

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -2,143 +2,180 @@
             {
-                // The user has specified an explicit root
-                return _relativeRootSolutionFolders;
+                var relativePathDirectory = Path.GetDirectoryName(pathRelativeToSolutionFile);
```
Yes. If the project file and the solution file are both in the same directory (aka, the project is not a nested project and does not need intermediate folders), then `pathRelativeToSolutionFile = "./Foo.csproj"` and therefore `relativePathDirectory = ""`

nagilson replied:
Ah I see what this is doing now, thank you. Looks good.

I feel that `relativeProjectPath` is a clearer name for this variable --`solutionRelativeProjectPath` is similarly good. Thats shorter than `projectPathRelativeToSolution` but that could work too. But I'm ok with keeping the current name. What do you (or others) think?

edvilme replied:
Yes, I agree. What do you think about `relativeDirectoryPath` and `solutionRelativePath`?

edvilme committed:
Nit: Update comments for readability

edvilme commented:
Hello @richardstanton whenever you have the chance, could you please take a quick look into this PR just so we can be sure we are not doing anything wrong with the API.
Also thank you so much for all your support during this process, greatly appreciate it!

nagilson reviewed (approved):
Excellent work here. I think you did a great job looking at the solution persistence code and implementing this. This will likely cause several breaking changes, which should be documented as mentioned in our comments above and also prior discussions with the SLN team. Also, thanks to @/richardstanton for doing some calls with us on this to determine the desired behavior, and @kasperk81 / others for helping!

Please do follow up with the remaining comments before merging -- and it would be ideal to get a review from the SLNX side, but not necessary. I'm approving because everything has been tentatively accounted for and the changes look good (don't want to block this for eternity) but it is important that the remaining items have some sort of resolution before merging. (Also, I will be gone soon and I want to give you a thumbs up while I'm still here.)

edvilme commented:
> This will likely cause several breaking changes, which should be documented as mentioned in our comments above and also prior discussions with the SLN team.

Yes, I will review those. For now, I think the breaking changes would be: 
- projects, solution folders, (and soon loose files) cannot longer have invalid names; 
- the command will no longer fail when attempting to add a nested project with the same name but different path;
- default configuration for solution files will be adding Any CPU, x86 and x64 platforms, and Debug and Release build types (I will try researching a bit more on that)

edvilme commented:
> Excellent work here. I think you did a great job looking at the solution persistence code and implementing this. This will likely cause several breaking changes, which should be documented as mentioned in our comments above and also prior discussions with the SLN team. Also, thanks to @/richardstanton for doing some calls with us on this to determine the desired behavior, and @kasperk81 / others for helping!

Thank you, and thanks for all your support ant the thorough review. Likewise thanks to @kasperk81 and @/richardstanton for all your support here :)

edvilme committed:
Remove unused code

edvilme committed:
NIT: Make default solution configuration constant

edvilme committed:
Merge branch 'main' into edvilme-slnx-add

edvilme committed:
[FIX] sln-add: Default solution folders for nested projects

edvilme committed:
Nit: Remove comment

edvilme commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -1,144 +1,180 @@
+                    {
+                        relativeSolutionFolder = Path.Combine(relativeSolutionFolder.Split(Path.DirectorySeparatorChar).SkipLast(1).ToArray());
+                    }
+                    if (!string.IsNullOrEmpty(relativeSolutionFolder))
```
Its value is changed in line 109, so it is possible that it becomes empty there

edvilme commented:
/backport to release/9.0.2xx

Bartleby2718 commented:
@edvilme @kasperk81 @nagilson @baronfel @Forgind I'm thinking of picking up https://github.com/dotnet/sdk/pull/45072 again, but I want to make sure we're aligned on the approach.

First, my personal preferences:
1. I prefer using giving descriptive variable names and using the latest syntax (e.g. collection expressions).
2. I prefer small, incremental PRs as they are easier to understand and review.

Problems for each:
1. These style changes can be distracting for reviewers.
2. A lot of PRs get merged to this repository, so merge conflicts (e.g. in `resx` files) can happen, which isn't the end of the world but can be annoying for me to resolve.

Do you want a big PR (like https://github.com/dotnet/sdk/pull/45072 currently is) that includes style changes, or do you want a few small PRs (perhaps with style/resx changes coming first)?

edvilme commented:
Hello @Bartleby2718 
Firstly thanks for your interest in this, I believe this will be very useful functionality. 

Regarding the PRs I agree that atomic PRs might be easier to review and develop than massive PRs. However my personal suggestion would be to find something in the middle and making sure those are independent, as to avoid having a lot of merge conflicts, backports and reviews on different PRs. 
I would recommend maybe creating different PRs `dotnet sln add folder`, and `dotnet sln add file` as they would be independent functionalities. 

As for the style, whatever works best for you, but due to the size of the repo, I would suggest attaching to the code conventions used as it makes it easier to maintain and contribute to in the future.

jurakovic commented:
Im on Windows, using Git Bash, dotnet version 9.0.200.

When I execute this:
```
dotnet sln Some.Solution.sln add ../Some.Project/src/Some.Project/Some.Project.csproj
```

I get this error
```
Names cannot:
- contain any of the following characters: / ? : \ * " < > |
- contain control characters
- be system reserved names, including 'CON', 'AUX', 'PRN', 'COM1' or 'LPT2'
- be '.' or '..' (Parameter 'name')
```

Is this expected behavior? Because using Visual Studio I can normally add project like that.

This worked fine until I have updated VS do 17.13.0 today.
I have a lot of solutions and projects and some scripts that automated process of adding projects to solutions, but all of that doesn't work now.

This is example from sln file after I have manually added project, but that manual work doesn't scale quite as bash scripts :)
```
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Some.Project", "..\Some.Project\src\Some.Project\Some.Project.csproj", "{4075164D-23CE-C114-6999-4E73B7467650}"
```

How can I add project with path starting with ".." using cli?

edvilme commented:
Hello @kasperk81 @jurakovic thanks for reaching out. Yes, we are aware of this bug when adding projects outside the solution file directory (and descendants) and already have a fix #46456 that should be released with 9.0.201

philipp-naused commented:
@jurakovic You can install multiple versions of the SDK in parallel.
<https://dotnet.microsoft.com/en-us/download/dotnet/9.0>
Add a `global.json` file to the root of your repo like this:
```json
{
  "sdk": {
    "version": "9.0.102",
    "rollForward": "disable"
  }
}
```
